crates/accelerate/src/quantum_circuit/circuit_data.rs

```diff
@@ -0,0 +1,387 @@
+            }
+            Int(index) => {
+                let index = self.convert_py_index(index, IndexFor::Lookup)?;
+                let extract_args =
+                    |bits: &PyList, args: &Vec<BitType>| -> PyResult<Vec<PyObject>> {
+                        args.iter()
+                            .map(|i| bits.get_item(*i as usize).map(|x| x.into()))
+                            .collect()
+                    };
+
+                if let Some(InternedInstruction(op, qargs_slot, cargs_slot)) = self.data.get(index)
+                {
+                    let context = self.context(py)?;
+                    let qargs = context.lookup(*qargs_slot);
+                    let cargs = context.lookup(*cargs_slot);
+                    self.new_callable.as_ref().unwrap().call1(
+                        py,
+                        (
+                            op,
+                            extract_args(self.qubits.as_ref(py), qargs)?,
+                            extract_args(self.clbits.as_ref(py), cargs)?,
+                        ),
+                    )
+                } else {
+                    Err(PyIndexError::new_err(format!(
+                        "No element at index {index} in circuit data",
+                    )))
+                }
+            }
+        }
+    }
+
+    pub fn __delitem__(&mut self, py: Python<'_>, index: SliceOrInt) -> PyResult<()> {
+        match index {
+            Slice(slice) => {
+                let slice = self.convert_py_slice(py, slice)?;
+                for i in slice {
+                    self.__delitem__(py, Int(i))?;
+                }
+                Ok(())
+            }
+            Int(index) => {
+                let index = self.convert_py_index(index, IndexFor::Lookup)?;
+                if self.data.get(index).is_some() {
+                    let cached_entry = self.data.remove(index);
+                    self.drop_from_cache(py, cached_entry)
+                } else {
+                    Err(PyIndexError::new_err(format!(
+                        "No element at index {index} in circuit data",
+                    )))
+                }
+            }
+        }
+    }
+
+    pub fn __setitem__(
+        &mut self,
+        py: Python<'_>,
+        index: SliceOrInt,
+        value: &PyAny,
+    ) -> PyResult<()> {
+        match index {
+            Slice(slice) => {
+                let indices = slice.indices(self.data.len().try_into().unwrap())?;
+                let slice = self.convert_py_slice(py, slice)?;
+                let values = value.iter()?.collect::<PyResult<Vec<_>>>()?;
+                if indices.step != 1 && slice.len() != values.len() {
+                    return Err(PyValueError::new_err(format!(
+                        "attempt to assign sequence of size {:?} to extended slice of size {:?}",
+                        values.len(),
+                        slice.len(),
+                    )));
+                }
```

Comment:
Added a comment in 1413711.

This particular special casing is here to support the use case of replacing a contiguous sublist of N items with another of length M. 

I'm not sure exactly if it invalidates the `ParameterTable`, but in general, setting with a slice is not actually possible behind the existing `QuantumCircuitData` wrapper anyway—its `__setitem__` implementation does not handle slices (which seems like it might be a minor bug, at least with respect to deprecation policy).

This _is_ used however directly via `QuantumCircuit._data` in `lin_comb.py` and maybe in some other places. I figured it'd make the most sense to support slicing fully in case we change `QuantumCircuit.data` to return a `CircuitData` directly without the `QuantumCircuitData` wrapper and still need to honor slicing for some deprecation period.